reference/reorder/mc64_kernels.cpp

```diff
@@ -0,0 +1,440 @@
+    auto idxs = index_workspace.get_data() + 4 * num_rows;
+    auto unmatched = idxs + num_rows;
+    auto um_cnt = 0;
+
+    // For each row, look for an unmatched column col for which weight(row, col)
+    // = 0. If one is found, add the edge (row, col) to the matching and move on
+    // to the next row.
+    for (IndexType row = 0; row < num_rows; row++) {
+        const auto row_begin = row_ptrs[row];
+        const auto row_end = row_ptrs[row + 1];
+        bool matched = false;
+        for (IndexType idx = row_begin; idx < row_end; idx++) {
+            const auto col = col_idxs[idx];
+            if (abs(weights[idx] - dual_u[col]) < tolerance && ip[col] == -1) {
+                p[row] = col;
+                ip[col] = row;
+                idxs[row] = idx;
+                matched = true;
+                break;
+            }
+        }
+        if (!matched) {
+            // Mark unmatched rows for later.
+            unmatched[um_cnt++] = row;
+        }
+    }
+
+    // For remaining unmatched rows, look for a matched column with weight(row,
+    // col) = 0 that is matched to another row, row_1. If there is another
+    // column col_1 with weight(row_1, col_1) = 0 that is not yet matched,
+    // replace the matched edge (row_1, col) with the two new matched edges
+    // (row, col) and (row_1, col_1).
+    auto um = 0;
+    auto row = unmatched[um];
+    // If row == 0 we passed the last unmatched row and reached the
+    // zero-initialized part of the array. Row 0 is always matched as the matrix
+    // is assumed to be nonsingular and the previous loop starts with row 0.
+    while (row != 0 && um < num_rows) {
+        const auto row_begin = row_ptrs[row];
+        const auto row_end = row_ptrs[row + 1];
+        bool found = false;
+        for (IndexType idx = row_begin; idx < row_end; idx++) {
+            const auto col = col_idxs[idx];
+            if (abs(weights[idx] - dual_u[col]) < tolerance) {
+                const auto row_1 = ip[col];
+                const auto row_1_begin = row_ptrs[row_1];
+                const auto row_1_end = row_ptrs[row_1 + 1];
+                for (IndexType idx_1 = row_1_begin; idx_1 < row_1_end;
+                     idx_1++) {
+                    const auto col_1 = col_idxs[idx_1];
+                    if (abs(weights[idx_1] - dual_u[col_1]) < tolerance &&
+                        ip[col_1] == -1) {
+                        p[row] = col;
+                        ip[col] = row;
+                        idxs[row] = idx;
+                        p[row_1] = col_1;
+                        ip[col_1] = row_1;
+                        idxs[row_1] = idx_1;
+                        found = true;
+                        break;
+                    }
+                }
+                if (found) break;
+            }
+        }
+        if (found) {
+            // Mark previously unmatched row as matched.
+            unmatched[um] = -1;
+        }
+        row = unmatched[++um];
```

Comment:
This will not happen: `unmatched` contains all indices of currently unmatched rows followed by as many zeros as already matched rows. It is processed from left to right. Once a row is marked as matched with `unmatched[row] = -1`, we move one to the right and process the next unmatched row, not returning to the previous row.